feat(audit-log): add common audit-log package

Zaperex opened the pull request:
### Description
Add initial common package for the audit logging
Fixes: [RHIDP-2287](https://issues.redhat.com/browse/RHIDP-2287)

schultzp2020 commented on plugins/audit-log-common/src/types.ts:
```diff
@@ -0,0 +1,37 @@
+};
+
+// TODO: Get proper type for body
+export type AuditRequest = {
+  body: unknown;
+  uri: string;
+  method: string;
+};
+export type AuditResponse = {
+  body: unknown;
+  status: number;
+};
+export type auditLog = {
+  actor: ActorDetails;
+  eventName: string;
+  stage: string;
+  request?: AuditRequest;
+  response?: AuditResponse;
+  meta: { [key: string]: any };
+  isAuditLog: true;
+} & auditStatus;
```
```suggestion
export type AuditResponse = {
  body: unknown;
  status: number;
};
export type auditLog = {
  actor: ActorDetails;
  eventName: string;
  stage: string;
  request?: Request;
  response?: Response;
  meta: { [key: string]: any };
  isAuditLog: true;
} & auditStatus;
```

Zaperex replied:
@schultzp2020 In the call I had with @AndrienkoAleksandr , we decided not to put the entire request/response object as to not expose tokens (from headers) and other irrelevant fields.

Zaperex replied:
For example, if I include the `Request` object in the audit log (note: the audit log structure is not the same as the one in this pull request), I get this extremely bloated audit log (I prettified it just to make it more readable):

```json
{
  "actor": {
    "client": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "ip_address": "::1",
    "user_id": "user:default/zaperex"
  },
  "event_name": "scaffolderTaskCreation",
  "isAuditLog": true,
  "level": "info",
  "message": "Scaffolding task for template:default/test-template with taskId: 3e7664d9-4400-4a66-825c-bde1ceaaf863 created by user:default/zaperex",
  "metadata": {
    "request": {
      "_body": true,
      "_consuming": false,
      "_dumped": false,
      "_events": {},
      "_eventsCount": 0,
      "_parsedUrl": {
        "_raw": "/v2/tasks",
        "auth": null,
        "hash": null,
        "host": null,
        "hostname": null,
        "href": "/v2/tasks",
        "path": "/v2/tasks",
        "pathname": "/v2/tasks",
        "port": null,
        "protocol": null,
        "query": null,
        "search": null,
        "slashes": null
      },
      "_readableState": {
        "autoDestroy": true,
        "awaitDrainWriters": null,
        "buffer": {
          "head": null,
          "length": 0,
          "tail": null
        },
        "closeEmitted": true,
        "closed": true,
        "constructed": true,
        "dataEmitted": true,
        "decoder": null,
        "defaultEncoding": "utf8",
        "destroyed": true,
        "emitClose": true,
        "emittedReadable": false,
        "encoding": null,
        "endEmitted": true,
        "ended": true,
        "errorEmitted": false,
        "errored": null,
        "flowing": true,
        "highWaterMark": 16384,
        "length": 0,
        "multiAwaitDrain": false,
        "needReadable": false,
        "objectMode": false,
        "pipes": [],
        "readableListening": false,
        "reading": false,
        "readingMore": true,
        "resumeScheduled": false,
        "sync": true
      },
      "_remoteAddress": "::1",
      "_startAt": [
        51443,
        772574557
      ],
      "_startTime": "2024-05-08T18:54:10.461Z",
      "aborted": false,
      "baseUrl": "/api/scaffolder",
      "body": {
        "secrets": {
          "password": "sfgdhjkl"
        },
        "templateRef": "template:default/test-template",
        "values": {
          "password": "********",
          "username": "program",
          "wait_time": 10
        }
      },
      "client": {
        "_closeAfterHandlingError": false,
        "_events": {
          "close": [
            null,
            null,
            null,
            null
          ],
          "end": [
            null,
            null
          ],
          "error": [
            null,
            null
          ]
        },
        "_eventsCount": 8,
        "_hadError": false,
        "_host": null,
        "_httpMessage": {
          "_closed": false,
          "_contentLength": null,
          "_defaultKeepAlive": true,
          "_events": {
            "finish": [
              null,
              null,
              null
            ]
          },
          "_eventsCount": 2,
          "_expect_continue": false,
          "_hasBody": true,
          "_header": null,
          "_headerSent": false,
          "_keepAliveTimeout": 5000,
          "_last": false,
          "_maxRequestsPerSocket": 0,
          "_removedConnection": false,
          "_removedContLen": false,
          "_removedTE": false,
          "_sent100": false,
          "_trailer": "",
          "chunkedEncoding": false,
          "destroyed": false,
          "finished": false,
          "locals": {},
          "maxRequestsOnConnectionReached": false,
          "outputData": [],
          "outputSize": 0,
          "req": "[Circular]",
          "sendDate": true,
          "shouldKeepAlive": true,
          "socket": "[Circular]",
          "strictContentLength": false,
          "useChunkedEncodingByDefault": true,
          "writable": true
        },
        "_parent": null,
        "_paused": false,
        "_peername": {
          "address": "::1",
          "family": "IPv6",
          "port": 59434
        },
        "_pendingData": null,
        "_pendingEncoding": "",
        "_readableState": {
          "autoDestroy": true,
          "awaitDrainWriters": null,
          "buffer": {
            "head": null,
            "length": 0,
            "tail": null
          },
          "closeEmitted": false,
          "closed": false,
          "constructed": true,
          "dataEmitted": false,
          "decoder": null,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emitClose": false,
          "emittedReadable": false,
          "encoding": null,
          "endEmitted": false,
          "ended": false,
          "errorEmitted": false,
          "errored": null,
          "flowing": true,
          "highWaterMark": 16384,
          "length": 0,
          "multiAwaitDrain": false,
          "needReadable": true,
          "objectMode": false,
          "pipes": [],
          "readableListening": false,
          "reading": true,
          "readingMore": false,
          "resumeScheduled": false,
          "sync": false
        },
        "_server": {
          "_connectionKey": "6::::7007",
          "_connections": 2,
          "_events": {
            "connection": [
              null,
              null
            ],
            "request": [
              null,
              null
            ]
          },
          "_eventsCount": 3,
          "_handle": {
            "reading": false
          },
          "_pendingSockets": {},
          "_unref": false,
          "_usingWorkers": false,
          "_workers": [],
          "allowHalfOpen": true,
          "connectionsCheckingInterval": 30000,
          "headersTimeout": 60000,
          "highWaterMark": 16384,
          "httpAllowHalfOpen": false,
          "keepAlive": false,
          "keepAliveInitialDelay": 0,
          "keepAliveTimeout": 5000,
          "maxHeadersCount": null,
          "maxRequestsPerSocket": 0,
          "noDelay": true,
          "pauseOnConnect": false,
          "rejectNonStandardBodyWrites": false,
          "requestTimeout": 300000,
          "timeout": 0
        },
        "_sockname": null,
        "_writableState": {
          "afterWriteTickInfo": null,
          "allBuffers": true,
          "allNoop": true,
          "autoDestroy": true,
          "bufferProcessing": false,
          "buffered": [],
          "bufferedIndex": 0,
          "closeEmitted": false,
          "closed": false,
          "constructed": true,
          "corked": 0,
          "decodeStrings": false,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emitClose": false,
          "ended": false,
          "ending": false,
          "errorEmitted": false,
          "errored": null,
          "finalCalled": false,
          "finished": false,
          "highWaterMark": 16384,
          "length": 0,
          "needDrain": false,
          "objectMode": false,
          "pendingcb": 0,
          "prefinished": false,
          "sync": false,
          "writecb": null,
          "writelen": 0,
          "writing": false
        },
        "allowHalfOpen": true,
        "connecting": false,
        "parser": {
          "0": null,
          "_consumed": true,
          "_headers": [],
          "_url": "",
          "incoming": "[Circular]",
          "joinDuplicateHeaders": null,
          "maxHeaderPairs": 2000,
          "outgoing": null,
          "socket": "[Circular]"
        },
        "server": {
          "_connectionKey": "6::::7007",
          "_connections": 2,
          "_events": {
            "connection": [
              null,
              null
            ],
            "request": [
              null,
              null
            ]
          },
          "_eventsCount": 3,
          "_handle": {
            "reading": false
          },
          "_pendingSockets": {},
          "_unref": false,
          "_usingWorkers": false,
          "_workers": [],
          "allowHalfOpen": true,
          "connectionsCheckingInterval": 30000,
          "headersTimeout": 60000,
          "highWaterMark": 16384,
          "httpAllowHalfOpen": false,
          "keepAlive": false,
          "keepAliveInitialDelay": 0,
          "keepAliveTimeout": 5000,
          "maxHeadersCount": null,
          "maxRequestsPerSocket": 0,
          "noDelay": true,
          "pauseOnConnect": false,
          "rejectNonStandardBodyWrites": false,
          "requestTimeout": 300000,
          "timeout": 0
        },
        "timeout": 0
      },
      "complete": true,
      "httpVersion": "1.1",
      "httpVersionMajor": 1,
      "httpVersionMinor": 1,
      "joinDuplicateHeaders": null,
      "method": "POST",
      "originalUrl": "/api/scaffolder/v2/tasks",
      "params": {},
      "query": {},
      "rawHeaders": [
        "Host",
        "localhost:7007",
        "Connection",
        "keep-alive",
        "Content-Length",
        "151",
        "sec-ch-ua",
        "\"Chromium\";v=\"124\", \"Google Chrome\";v=\"124\", \"Not-A.Brand\";v=\"99\"",
        "Content-Type",
        "application/json",
        "sec-ch-ua-mobile",
        "?0",
        "authorization",
        "Bearer eyJ0eXAiOiJ2bmQuYmFja3N0YWdlLnVzZXIiLCJhbGciOiJFUzI1NiIsImtpZCI6ImY1N2Y2NGFlLTA0MzYtNDIyMC1hYWYxLWE0NDMyNDI0Mjg5MSJ9.eyJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjcwMDcvYXBpL2F1dGgiLCJzdWIiOiJ1c2VyOmRlZmF1bHQvemFwZXJleCIsImVudCI6WyJ1c2VyOmRlZmF1bHQvemFwZXJleCJdLCJhdWQiOiJiYWNrc3RhZ2UiLCJpYXQiOjE3MTUxOTQ0MzgsImV4cCI6MTcxNTE5ODAzOCwidWlwIjoiR3ZKWmo4bC0tdWVGTUhSV0dkVHNUY1dIVURmeEdWMEZEQmdtNEJRYkpvQllMOWdkSERvVWtmLWRWanhqLWI5VERvNndOWmMtcHItLWtqU1IxVV9BWkEifQ.sZ9ByvNsweIRi1z1HUmw1NljpiiauBvri0JL-f53yAB9UNCdscUpr0pZFhftPLhB7frfh0Gu1iTG35jZI28LPg",
        "User-Agent",
        "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
        "sec-ch-ua-platform",
        "\"Linux\"",
        "Accept",
        "*/*",
        "Origin",
        "http://localhost:3000",
        "Sec-Fetch-Site",
        "same-site",
        "Sec-Fetch-Mode",
        "cors",
        "Sec-Fetch-Dest",
        "empty",
        "Referer",
        "http://localhost:3000/",
        "Accept-Encoding",
        "gzip, deflate, br, zstd",
        "Accept-Language",
        "en-US,en;q=0.9"
      ],
      "rawTrailers": [],
      "res": {
        "_closed": false,
        "_contentLength": null,
        "_defaultKeepAlive": true,
        "_events": {
          "finish": [
            null,
            null,
            null
          ]
        },
        "_eventsCount": 2,
        "_expect_continue": false,
        "_hasBody": true,
        "_header": null,
        "_headerSent": false,
        "_keepAliveTimeout": 5000,
        "_last": false,
        "_maxRequestsPerSocket": 0,
        "_removedConnection": false,
        "_removedContLen": false,
        "_removedTE": false,
        "_sent100": false,
        "_trailer": "",
        "chunkedEncoding": false,
        "destroyed": false,
        "finished": false,
        "locals": {},
        "maxRequestsOnConnectionReached": false,
        "outputData": [],
        "outputSize": 0,
        "req": "[Circular]",
        "sendDate": true,
        "shouldKeepAlive": true,
        "socket": {
          "_closeAfterHandlingError": false,
          "_events": {
            "close": [
              null,
              null,
              null,
              null
            ],
            "end": [
              null,
              null
            ],
            "error": [
              null,
              null
            ]
          },
          "_eventsCount": 8,
          "_hadError": false,
          "_host": null,
          "_httpMessage": "[Circular]",
          "_parent": null,
          "_paused": false,
          "_peername": {
            "address": "::1",
            "family": "IPv6",
            "port": 59434
          },
          "_pendingData": null,
          "_pendingEncoding": "",
          "_readableState": {
            "autoDestroy": true,
            "awaitDrainWriters": null,
            "buffer": {
              "head": null,
              "length": 0,
              "tail": null
            },
            "closeEmitted": false,
            "closed": false,
            "constructed": true,
            "dataEmitted": false,
            "decoder": null,
            "defaultEncoding": "utf8",
            "destroyed": false,
            "emitClose": false,
            "emittedReadable": false,
            "encoding": null,
            "endEmitted": false,
            "ended": false,
            "errorEmitted": false,
            "errored": null,
            "flowing": true,
            "highWaterMark": 16384,
            "length": 0,
            "multiAwaitDrain": false,
            "needReadable": true,
            "objectMode": false,
            "pipes": [],
            "readableListening": false,
            "reading": true,
            "readingMore": false,
            "resumeScheduled": false,
            "sync": false
          },
          "_server": {
            "_connectionKey": "6::::7007",
            "_connections": 2,
            "_events": {
              "connection": [
                null,
                null
              ],
              "request": [
                null,
                null
              ]
            },
            "_eventsCount": 3,
            "_handle": {
              "reading": false
            },
            "_pendingSockets": {},
            "_unref": false,
            "_usingWorkers": false,
            "_workers": [],
            "allowHalfOpen": true,
            "connectionsCheckingInterval": 30000,
            "headersTimeout": 60000,
            "highWaterMark": 16384,
            "httpAllowHalfOpen": false,
            "keepAlive": false,
            "keepAliveInitialDelay": 0,
            "keepAliveTimeout": 5000,
            "maxHeadersCount": null,
            "maxRequestsPerSocket": 0,
            "noDelay": true,
            "pauseOnConnect": false,
            "rejectNonStandardBodyWrites": false,
            "requestTimeout": 300000,
            "timeout": 0
          },
          "_sockname": null,
          "_writableState": {
            "afterWriteTickInfo": null,
            "allBuffers": true,
            "allNoop": true,
            "autoDestroy": true,
            "bufferProcessing": false,
            "buffered": [],
            "bufferedIndex": 0,
            "closeEmitted": false,
            "closed": false,
            "constructed": true,
            "corked": 0,
            "decodeStrings": false,
            "defaultEncoding": "utf8",
            "destroyed": false,
            "emitClose": false,
            "ended": false,
            "ending": false,
            "errorEmitted": false,
            "errored": null,
            "finalCalled": false,
            "finished": false,
            "highWaterMark": 16384,
            "length": 0,
            "needDrain": false,
            "objectMode": false,
            "pendingcb": 0,
            "prefinished": false,
            "sync": false,
            "writecb": null,
            "writelen": 0,
            "writing": false
          },
          "allowHalfOpen": true,
          "connecting": false,
          "parser": {
            "0": null,
            "_consumed": true,
            "_headers": [],
            "_url": "",
            "incoming": "[Circular]",
            "joinDuplicateHeaders": null,
            "maxHeaderPairs": 2000,
            "outgoing": null,
            "socket": "[Circular]"
          },
          "server": {
            "_connectionKey": "6::::7007",
            "_connections": 2,
            "_events": {
              "connection": [
                null,
                null
              ],
              "request": [
                null,
                null
              ]
            },
            "_eventsCount": 3,
            "_handle": {
              "reading": false
            },
            "_pendingSockets": {},
            "_unref": false,
            "_usingWorkers": false,
            "_workers": [],
            "allowHalfOpen": true,
            "connectionsCheckingInterval": 30000,
            "headersTimeout": 60000,
            "highWaterMark": 16384,
            "httpAllowHalfOpen": false,
            "keepAlive": false,
            "keepAliveInitialDelay": 0,
            "keepAliveTimeout": 5000,
            "maxHeadersCount": null,
            "maxRequestsPerSocket": 0,
            "noDelay": true,
            "pauseOnConnect": false,
            "rejectNonStandardBodyWrites": false,
            "requestTimeout": 300000,
            "timeout": 0
          },
          "timeout": 0
        },
        "strictContentLength": false,
        "useChunkedEncodingByDefault": true,
        "writable": true
      },
      "route": {
        "methods": {
          "post": true
        },
        "path": "/v2/tasks",
        "stack": [
          {
            "keys": [],
            "method": "post",
            "name": "<anonymous>",
            "regexp": {
              "fast_slash": false,
              "fast_star": false
            }
          }
        ]
      },
      "socket": {
        "_closeAfterHandlingError": false,
        "_events": {
          "close": [
            null,
            null,
            null,
            null
          ],
          "end": [
            null,
            null
          ],
          "error": [
            null,
            null
          ]
        },
        "_eventsCount": 8,
        "_hadError": false,
        "_host": null,
        "_httpMessage": {
          "_closed": false,
          "_contentLength": null,
          "_defaultKeepAlive": true,
          "_events": {
            "finish": [
              null,
              null,
              null
            ]
          },
          "_eventsCount": 2,
          "_expect_continue": false,
          "_hasBody": true,
          "_header": null,
          "_headerSent": false,
          "_keepAliveTimeout": 5000,
          "_last": false,
          "_maxRequestsPerSocket": 0,
          "_removedConnection": false,
          "_removedContLen": false,
          "_removedTE": false,
          "_sent100": false,
          "_trailer": "",
          "chunkedEncoding": false,
          "destroyed": false,
          "finished": false,
          "locals": {},
          "maxRequestsOnConnectionReached": false,
          "outputData": [],
          "outputSize": 0,
          "req": "[Circular]",
          "sendDate": true,
          "shouldKeepAlive": true,
          "socket": "[Circular]",
          "strictContentLength": false,
          "useChunkedEncodingByDefault": true,
          "writable": true
        },
        "_parent": null,
        "_paused": false,
        "_peername": {
          "address": "::1",
          "family": "IPv6",
          "port": 59434
        },
        "_pendingData": null,
        "_pendingEncoding": "",
        "_readableState": {
          "autoDestroy": true,
          "awaitDrainWriters": null,
          "buffer": {
            "head": null,
            "length": 0,
            "tail": null
          },
          "closeEmitted": false,
          "closed": false,
          "constructed": true,
          "dataEmitted": false,
          "decoder": null,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emitClose": false,
          "emittedReadable": false,
          "encoding": null,
          "endEmitted": false,
          "ended": false,
          "errorEmitted": false,
          "errored": null,
          "flowing": true,
          "highWaterMark": 16384,
          "length": 0,
          "multiAwaitDrain": false,
          "needReadable": true,
          "objectMode": false,
          "pipes": [],
          "readableListening": false,
          "reading": true,
          "readingMore": false,
          "resumeScheduled": false,
          "sync": false
        },
        "_server": {
          "_connectionKey": "6::::7007",
          "_connections": 2,
          "_events": {
            "connection": [
              null,
              null
            ],
            "request": [
              null,
              null
            ]
          },
          "_eventsCount": 3,
          "_handle": {
            "reading": false
          },
          "_pendingSockets": {},
          "_unref": false,
          "_usingWorkers": false,
          "_workers": [],
          "allowHalfOpen": true,
          "connectionsCheckingInterval": 30000,
          "headersTimeout": 60000,
          "highWaterMark": 16384,
          "httpAllowHalfOpen": false,
          "keepAlive": false,
          "keepAliveInitialDelay": 0,
          "keepAliveTimeout": 5000,
          "maxHeadersCount": null,
          "maxRequestsPerSocket": 0,
          "noDelay": true,
          "pauseOnConnect": false,
          "rejectNonStandardBodyWrites": false,
          "requestTimeout": 300000,
          "timeout": 0
        },
        "_sockname": null,
        "_writableState": {
          "afterWriteTickInfo": null,
          "allBuffers": true,
          "allNoop": true,
          "autoDestroy": true,
          "bufferProcessing": false,
          "buffered": [],
          "bufferedIndex": 0,
          "closeEmitted": false,
          "closed": false,
          "constructed": true,
          "corked": 0,
          "decodeStrings": false,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emitClose": false,
          "ended": false,
          "ending": false,
          "errorEmitted": false,
          "errored": null,
          "finalCalled": false,
          "finished": false,
          "highWaterMark": 16384,
          "length": 0,
          "needDrain": false,
          "objectMode": false,
          "pendingcb": 0,
          "prefinished": false,
          "sync": false,
          "writecb": null,
          "writelen": 0,
          "writing": false
        },
        "allowHalfOpen": true,
        "connecting": false,
        "parser": {
          "0": null,
          "_consumed": true,
          "_headers": [],
          "_url": "",
          "incoming": "[Circular]",
          "joinDuplicateHeaders": null,
          "maxHeaderPairs": 2000,
          "outgoing": null,
          "socket": "[Circular]"
        },
        "server": {
          "_connectionKey": "6::::7007",
          "_connections": 2,
          "_events": {
            "connection": [
              null,
              null
            ],
            "request": [
              null,
              null
            ]
          },
          "_eventsCount": 3,
          "_handle": {
            "reading": false
          },
          "_pendingSockets": {},
          "_unref": false,
          "_usingWorkers": false,
          "_workers": [],
          "allowHalfOpen": true,
          "connectionsCheckingInterval": 30000,
          "headersTimeout": 60000,
          "highWaterMark": 16384,
          "httpAllowHalfOpen": false,
          "keepAlive": false,
          "keepAliveInitialDelay": 0,
          "keepAliveTimeout": 5000,
          "maxHeadersCount": null,
          "maxRequestsPerSocket": 0,
          "noDelay": true,
          "pauseOnConnect": false,
          "rejectNonStandardBodyWrites": false,
          "requestTimeout": 300000,
          "timeout": 0
        },
        "timeout": 0
      },
      "statusCode": null,
      "statusMessage": null,
      "upgrade": false,
      "url": "/v2/tasks"
    },
    "taskId": "3e7664d9-4400-4a66-825c-bde1ceaaf863",
    "templateRef": "template:default/test-template"
  },
  "plugin": "scaffolder",
  "service": "backstage",
  "status": "success",
  "timestamp": "2024-05-08T18:54:10.480Z",
  "type": "plugin"
}
```

Zaperex replied:
Hmm, just noticed that the secrets inputted into the scaffolder software template was set in clear text in the body object. Need to figure out a way to redact it or maybe just not include the body at all?

AndrienkoAleksandr replied:
@schultzp2020, Apologies, but the Kubernetes (k8s) audit log doesn't track all requests and responses – they are indeed quite large. The k8s audit log only records request/response bodies. The audit log file will grow very quickly if we include the entire request/response. I anticipate that the audit log streamer will fail to write data when there is no more free disk space available.

AndrienkoAleksandr replied:
@Zaperex , I agree with you request/response body could contain sensitive information. I think that's why k8s audit log has level feature. Level values:

```
None - don't log events that match this rule.
Metadata - log request metadata (requesting user, timestamp, resource, verb, etc.) but not request or response body.
Request - log event metadata and request body but not response body. This does not apply for non-resource requests.
RequestResponse - log event metadata, request and response bodies. This does not apply for non-resource requests.
```

@schultzp2020 , @Zaperex Since we don't have strong agreement about request/response body, we can postpone this audit log feature and apply it in the next iteration. We need to consider how to hide sensitive information and we have a lack time for RBAC audit log. And maybe it make sense to have request/response body audit log message like optional feature. We can introduce auditLog application configuration and provide the way how enable/disable request/response body audit log messages. Each plugin with audit log implementation can use this configuration.



@Zaperex, I agree with your assessment that the request/response body could contain sensitive information. I believe this is why the k8s audit log incorporates a level feature. The level values are as follows:

```
None - don't log events that match this rule.
Metadata - log request metadata (requesting user, timestamp, resource, verb, etc.) but not the request or response body.
Request - log event metadata and the request body but not the response body. This does not apply to non-resource requests.
RequestResponse - log event metadata, the request, and response bodies. This does not apply to non-resource requests.
```

Admins can configure the system to track only Metadata audit log lines, without including the request/response.

@schultzp2020 , @Zaperex Since given the lack of strong consensus regarding the handling of request/response bodies, I suggest postponing this audit log feature and implementing it in the next iteration. We need to carefully consider how to hide sensitive information, and currently, our priority lies with RBAC audit logs due to time constraints. Additionally, it might be sensible to make the request/response body audit log message an optional feature. We can introduce an auditLog application configuration and provide a way to enable/disable request/response body audit log messages. Each plugin with audit log implementation can utilize this configuration.

So admin can configure to track only Metadata audit log lines, without request/response audit log lines.

Zaperex replied:
For now, I just added a comment for users to redact possibly sensitive information in the `req.body` before sending it to be audit logged.

AndrienkoAleksandr commented on plugins/audit-log-common/src/types.ts:
```diff
@@ -0,0 +1,37 @@
+  body: unknown;
+  status: number;
+};
+export type auditLog = {
```
```suggestion
export type AuditLog = {
```

Zaperex commented:
Added a test unit test in https://github.com/janus-idp/backstage-plugins/pull/1622/commits/5f8b45510577eefbfb47a365be3937874d31f09a, but for some reason the jest tests are completely borked with the following error: https://github.com/janus-idp/backstage-plugins/actions/runs/9069106978/job/24917922518#step:3:713 

I tried creating a new local common package in `backstage-plugins` and added a simple test like the following where `returnTrue` is just a function that returns `true`:
```ts
import { returnTrue } from "./randomThing";
import { getVoidLogger } from '@backstage/backend-common';


describe('Return True', () => {
  const logger = getVoidLogger();
  it ("returns true", () => {
    logger.info("Hello");
    expect(returnTrue).toBeTruthy();
  })
})
```
This works if I remove the `getVoidLogger`, but as soon as it's added back in, it fails with:
```ts
yarn run v1.22.21
$ backstage-cli package test --passWithNoTests --coverage
 FAIL  src/randomThing.test.ts
  ● Test suite failed to run

    ReferenceError: TextEncoder is not defined

      4 |
      5 | describe('Return True', () => {
    > 6 |   const logger = getVoidLogger();
        |                        ^
      7 |   it ("returns true", () => {
      8 |     logger.info("Hello");
      9 |     expect(returnTrue).toBeTruthy();

      at Object.<anonymous> (../../../node_modules/jose/dist/browser/lib/buffer_utils.js:2:24)
      at Object.<anonymous> (../../../node_modules/jose/dist/browser/runtime/base64url.js:25:23)
      at Object.<anonymous> (../../../node_modules/jose/dist/browser/jwe/flattened/decrypt.js:11:20)
      at Object.<anonymous> (../../../node_modules/jose/dist/browser/jwe/compact/decrypt.js:11:18)
      at Object.<anonymous> (../../../node_modules/jose/dist/browser/index.js:121:18)
      at Object.<anonymous> (../../../node_modules/@backstage/backend-app-api/dist/index.cjs.js:26:12)
      at Object.<anonymous> (../../../node_modules/@backstage/backend-common/dist/index.cjs.js:8:21)
      at Object.<anonymous> (randomThing.test.ts:6:24)
```

Zaperex commented:
After doing some testing, I realized that with the current implementation a `No actor id was provided for audit log` error would be thrown if you try to access an unauthenticated endpoint utilizing the audit log package without providing an explicit `actor_id` or request with a valid backstage credential to the audit log. 

This would not be ideal for local development/testing or if the end user wants to disable the default auth policy for some reason, as those endpoints would be rendered useless. The changes in https://github.com/janus-idp/backstage-plugins/pull/1622/commits/2c723a3905668db315d86b78da48b7d696265177 removes these checks and make it so that audit logs for unauthenticated endpoints will have an `actorId` of `null` to clearly indicate unauthenticated audit logs (at this point these would function more as debug logs). 

Do you think this makes sense to add?

Zaperex commented:
Added `const { TextEncoder, TextDecoder } = require("util");` to the `node_modules/jose/dist/browser/lib/buffer_utils.js` file according to this [stackoverflow comment](https://stackoverflow.com/questions/19697858/referenceerror-textencoder-is-not-defined#:~:text=add%20this%20line%20at%20top%20const%20%7B%20TextEncoder%2C%20TextDecoder%20%7D%20%3D%20require(%22util%22)%3B) which got rid of the error, but another module not found error is occuring:
```ts
yarn run v1.22.21
$ backstage-cli package test --passWithNoTests --coverage
 FAIL  src/DefaultAuditLogger.test.ts
  ● Test suite failed to run

    Cannot find module 'textextensions' from '../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js'

    Require stack:
      /home/frkong/coding/janus-idp/backstage-plugins/node_modules/@backstage/backend-test-utils/dist/index.cjs.js
      DefaultAuditLogger.test.ts

      4 | import { DefaultAuditLogger } from './DefaultAuditLogger';
      5 |
    > 6 | describe('DefaultAuditLogger', () => {
        |                           ^
      7 |   const logger = getVoidLogger();
      8 |
      9 |   const auditLogger = new DefaultAuditLogger({

      at Resolver._throwModNotFoundError (../../../node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js:50:22)
      at Object.<anonymous> (DefaultAuditLogger.test.ts:6:27)
```

Zaperex commented:
Tested with a new `plugin-backend` from `yarn new` and it seemed to work there? Not sure why a new `plugin-common` has these issues.

schultzp2020 commented:
> Tested with a new `plugin-backend` from `yarn new` and it seemed to work there? Not sure why a new `plugin-common` has these issues.

Maybe `plugin-common` is for the frontend?

Zaperex commented:
Ok it seems I most likely should have made a `plugin-node` plugin. Converting existing common package to a node package.

kim-tsao commented on plugins/audit-log-node/package.json:
```diff
@@ -0,0 +1,53 @@
+{
```
Can you update the [team/rhdh renovate rule](https://github.com/janus-idp/backstage-plugins/blob/main/.github/renovate.json#L191) to add this new plugin so it gets scanned?  It'll save me the trouble of opening a new PR :pray:

e.g.

```

      "matchFileNames": [
        "plugins/audit-log-node*/**",   
        ...
      ],
      "additionalBranchPrefix": "rhdh ",
      "addLabels": [
        "team/rhdh"
      ]
```

Zaperex replied:
Will do

Zaperex commented on plugins/audit-log-node/README.md:
```diff
@@ -0,0 +1,3 @@
+# @janus-idp/backstage-plugin-audit-log-node
+
+This package contains common types and utility functions for audit logging the backend
```
Yep currently updating the `README.md` with some usage instructions

Zaperex commented on plugins/audit-log-node/src/DefaultAuditLogger.test.ts:
```diff
@@ -0,0 +1,296 @@
+  });
+
+  describe('createAuditLogDetails', () => {
+    it('Generates an audit log meta object with a null actor if neither actor id or request is provided', async () => {
```
So the main reasoning is that there's currently nothing catching the error that used to be thrown which causes whatever the audit log was supposed to be audit logging to fail. This mainly would be an issue if the user decides to turn off the auth policy and permissions to do testing or something.

kim-tsao replied:
ok, so `null` would be an indicator that it's in test vs prod mode.  If `null` is encountered in prod then it's likely an error on our side.  Can we document this behaviour in the docs?

Zaperex replied:
yep I added it in the `README.md`: https://github.com/janus-idp/backstage-plugins/pull/1622/files#diff-918b0834ae33ff3b25b34cd124f87cbdfb13144cc3126edcbae22f3368160f0cR25
https://github.com/janus-idp/backstage-plugins/blob/3af03877acd31f9640efb80e2d6113b25ee14c51/plugins/audit-log-node/README.md?plain=1#L21-L25

schultzp2020 commented on plugins/audit-log-node/src/types.ts:
```diff
@@ -0,0 +1,113 @@
+  stage: string;
+  metadata?: JsonValue;
+  response?: AuditResponse;
+  actor_id?: string;
```
```suggestion
  actorId?: string;
```

schultzp2020 commented on plugins/audit-log-node/src/types.ts:
```diff
@@ -0,0 +1,113 @@
+  eventName: string;
+  message: string;
+  stage: string;
+  actor_id?: string;
```
```suggestion
  actorId?: string;
```

AndrienkoAleksandr commented on plugins/audit-log-node/package.json:
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@janus-idp/backstage-plugin-audit-log-node",
```
Maybe it can be  @janus-idp/backstage-plugin-audit-log-common?

schultzp2020 replied:
It can't because it needs to follow the backstage naming scheme.

AndrienkoAleksandr commented on plugins/audit-log-node/src/DefaultAuditLogger.ts:
```diff
@@ -0,0 +1,130 @@
+      return undefined;
+    }
+  }
+  async createAuditLogDetails(options: AuditLogDetailsOptions) {
```
Add return type, please.

AndrienkoAleksandr replied:
Then, I guess you will need to cast `auditLogDetails as JsonObject)` in the bottom of the code:

`this.logger.info(options.message, auditLogDetails as JsonObject);`

`this.logger.error(options.message, auditLogDetails as JsonObject);`

Zaperex replied:
Oh sorry, I forgot to port the return type in the original function over. It should be returning a `Promise<AuditLogDetails>` type.

AndrienkoAleksandr commented on plugins/audit-log-node/src/DefaultAuditLogger.ts:
```diff
@@ -0,0 +1,130 @@
+        }
+      : undefined;
+
+    const auditLogCommonDetails = {
```
For the RBAC backend plugin, some parts of the changes come from the policy file or the plugin itself, so we do not always have the IP address, hostname, or user agent. Therefore, I propose changing this code to:
  
```
    let actor: ActorDetails = { actorId };
    if (request) {
      actor.ip = request.ip,
      actor.hostname = request.hostname,
      actor.userAgent = request.get('user-agent')
    }
    
    const auditLogCommonDetails = {
      actor,
      meta: metadata || {},
      request: auditRequest,
      isAuditLog: true as const,
      response,
      eventName,
      stage,
    };
  ```
  
Then, this code will simply skip these fields when the request is absent, instead of printing undefined values.

AndrienkoAleksandr replied:
Hmm. I guess it would be nice to do the same with request and response:

```
    let actor: ActorDetails = { actorId };
    if (request) {
      actor.ip = request.ip;
      actor.hostname = request.hostname;
      actor.userAgent = request.get('user-agent');
    }

    const auditLogCommonDetails: AuditLogDetails = {
      actor,
      meta: metadata || {},
      isAuditLog: true as const,
      eventName,
      stage,
      status: 'succeeded',
    };

    if (auditRequest) {
      auditLogCommonDetails.request = auditRequest;
    }
    if (response) {
      auditLogCommonDetails.response = response;
    }

    if (status === 'failed') {
      const errs = options.errors as ErrorLike[];
      return {
        ...auditLogCommonDetails,
        status,
        errors: errs.map(err => {
          return {
            name: err.name,
            message: err.message,
            stack: err.stack,
          };
        }),
      };
    }

    return {
      ...auditLogCommonDetails,
    };
   ```

Zaperex replied:
I mainly kept the `undefined` printing since keys with `undefined` values don't actually get printed. I can make those changes to make it more verbose as well.

schultzp2020 replied:
@kim-tsao, would customers prefer a field to be null or to not exist at all (undefined)? I believe they might prefer undefined, as it would take up less storage.

kim-tsao replied:
@AndrienkoAleksandr can you clarify this scenario?

> For the RBAC backend plugin, some parts of the changes come from the policy file or the plugin itself,

Is this the case where a cluster admin makes a change to the config file?  If so, they should enable audit logging on the cluster to capture these events.  Generating audit logs without any of the required actor identifiers would not be useful at all.  Can we can suppress audit log generation as long as we are certain this is the _only_ scenario where actor identifiers cannot be obtained?

We will need to document that Customers need to turn on audit logging in the clusters if they are making changes there.

PatAKnight replied:
Yes, @AndrienkoAleksandr is referencing whenever a cluster admin makes changes to the csv file in the cluster, those changes are picked up by the pod, and then our plugin pulls them in.

Zaperex replied:
In my case for the scaffolder-backend, many of the actions are being done asynchronously by the backend, so I just manually set an `actorId` of `scaffolder-backend`.

kim-tsao replied:
@Zaperex I thought about having a generic ID but I was a bit concerned that we could be generating too many events.  Maybe this won't be a problem since configmap modifications shouldn't be too frequent? I guess it doesn't hurt to generate additional records with the generic ID as long as it's not too excessive.   We can consider generic ID events as "backend triggered" which customers can then use to investigate further with their cluster audit logs

Zaperex replied:
@kim-tsao I'm also going to point out that a good portion of the `scaffolder-backend` audit logging has a generic actorID since I'm audit logging each scaffolder task (and all it's steps/iterations), but it only logs when tasks are being executed so it will not always be on. Will there be any concerns there?

However, the amount of generic actorID events being generated by the `catalog-backend` will definitely be alot as the catalog will be constantly asynchronously making updates to the entities without any user interactions, so we might need to discuss what details we want to audit log or perhaps log some of the audit logs to a `debug` log level or something? Ex: this catalog entity was updated from a scheduled refresh. What do you think?

kim-tsao replied:
@Zaperex 

> I'm also going to point out that a good portion of the scaffolder-backend audit logging has a generic actorID since I'm audit logging each scaffolder task (and all it's steps/iterations), but it only logs when tasks are being executed so it will not always be on. Will there be any concerns there?

Can these generic audit logs be tied back to the software template (id perhaps?) execution that was initiated by the user?  

> However, the amount of generic actorID events being generated by the catalog-backend will definitely be alot as the catalog will be constantly asynchronously making updates to the entities without any user interactions, so we might need to discuss what details we want to audit log or perhaps log some of the audit logs to a debug log level or something? Ex: this catalog entity was updated from a scheduled refresh. What do you think?

This should be ok as long as it was a user interaction that caused the async update (I think we talked about having a unique identifier tie these two events together at one point in time).  If it's a service sync refresh, we can consider that out of scope.

Zaperex replied:
@kim-tsao 
> Can these generic audit logs be tied back to the software template (id perhaps?) execution that was initiated by the user?

Yep they can, most of the scaffolder related audit logs have a `taskId` metadata field for the corresponding task. There would also be an audit log specifying which user created a task with that `taskId`. It's also stored in the scaffolder database alongside the tasks. I was mainly asking if there would be concerns with the volume of events being audit logged with the scaffolder and catalog since you said `I thought about having a generic ID but I was a bit concerned that we could be generating too many events`

> This should be ok as long as it was a user interaction that caused the async update (I think we talked about having a unique identifier tie these two events together at one point in time). If it's a service sync refresh, we can consider that out of scope.

Pretty much all the changes made to the catalog are caused by these catalog service sync refresh (a refresh essentially is an entity getting processed/re-processed), so it's quite difficult to differentiate between ones triggered by explicitly caused by user interactions and ones caused by the backend since everything is added to the same processing queue. 

Ex: Suppose the `catalog-info.yaml` that the location entity points to changed, then the corresponding entity will be updated the next time it gets queued up to be processed. However, a user manually triggering a refresh would cause the same process to occur with the only difference being that the user put the entity/location higher up in the queue.

Zaperex commented on plugins/audit-log-node/src/DefaultAuditLogger.test.ts:
```diff
@@ -0,0 +1,296 @@
+          },
+        },
+      });
+      console.warn(JSON.stringify(auditLogDetails, null, 2));
```
No we don't, that was just a test console log I forgot to remove :sweat:

schultzp2020 commented on plugins/audit-log-node/src/types.ts:
```diff
@@ -0,0 +1,113 @@
+  response?: AuditResponse;
+  actorId?: string;
+  request?: Request;
+} & ({ status: 'succeeded' } | { status: 'failed'; errors: unknown[] });
```
You could reuse the status types from above here.

Zaperex replied:
Ah in this case, I use `unknown[]` instead of `ErrorLike[]` so I couldn't reuse it.

Zaperex commented on plugins/audit-log-node/src/DefaultAuditLogger.ts:
```diff
@@ -0,0 +1,143 @@
+      response: options.response,
+      metadata: options.metadata,
+    });
+    this.logger.info(options.message, auditLogDetails as JsonObject);
```
I can extend this to support logging to other levels as well?

PatAKnight reviewed (changes requested):
Was able to test this using the RBAC Backend plugin. Picked one of the endpoints and it worked as expected.

Just had a couple of small requests.

PatAKnight commented:
/lgtm

Going to wait to add approval until all of the conversations are resolved.

Zaperex commented:
Discussed with @PatAKnight and it was decided that it was fine to just keep it merged and address any issues that may occur in a separate PR.

schultzp2020 commented on plugins/audit-log-node/README.md:
````diff
@@ -0,0 +1,210 @@
+In the following example, we add a simple audit log for the `/health` endpoint of a plugin's router.
+
+```ts plugins/test/src/service/router.ts
+/* highlight-add-start */
+
+/* highlight-add-end */
````
```suggestion
```

schultzp2020 commented on plugins/audit-log-node/README.md:
````diff
@@ -0,0 +1,210 @@
+In the following example, we utilize the `auditErrorLog` utility function to generate and output an error log:
+
+```ts plugins/test/src/service/router.ts
+/* highlight-add-start */
+
+/* highlight-add-end */
````
```suggestion
```

schultzp2020 commented on plugins/audit-log-node/src/DefaultAuditLogger.ts:
```diff
@@ -0,0 +1,159 @@
+      response: options.response,
+      metadata: options.metadata,
+    });
+    switch (logLevel) {
+      case 'info':
+        this.logger.info(options.message, auditLogDetails as JsonObject);
+        return;
+      case 'debug':
+        this.logger.debug(options.message, auditLogDetails as JsonObject);
+        return;
+      case 'warn':
+        this.logger.warn(options.message, auditLogDetails as JsonObject);
+        return;
+      case 'error':
+        this.logger.error(options.message, auditLogDetails as JsonObject);
+        return;
+      default:
+        throw new Error(`Log level of ${logLevel} is not supported`);
+    }
```
```suggestion
    this.logger.log(logLevel, options.message, auditLogDetails as JsonObject)
```

Zaperex replied:
The backstage `LoggerService` does not have the `log` function unforunately.

Zaperex replied:
https://github.com/backstage/backstage/blob/b1ecb9f32536a279fd798a58d66e333ea3c4b1f0/packages/backend-plugin-api/src/services/definitions/LoggerService.ts#L24

schultzp2020 commented on plugins/audit-log-node/src/DefaultAuditLogger.ts:
```diff
@@ -0,0 +1,159 @@
+  async auditErrorLog(options: AuditErrorLogOptions): Promise<void> {
+    const auditLogDetails = await this.createAuditLogDetails({
+      eventName: options.eventName,
+      status: 'failed',
```
Should we hardcode the status in both implementations of the auditLog, or should we require the user to specify whether it was a success or failure? I believe we should make the user specify the outcome for less code duplication.

Zaperex replied:
I guess that could make sense to give more control to the user. We can just set the default statuses to `failed` and `succeeded` for the `auditErrorLog` and `auditLog` functions respectively if no status is supplied.

Zaperex replied:
Removed the `auditErrorLog` function in https://github.com/janus-idp/backstage-plugins/pull/1730/commits/35eb36afdbce0cb4e8fb5434d00c0e92a1edd7cc in favor of using the `auditLog` function for everything for simplicity. 

This would require the following changes:
- All usages of `auditErrorLog` to be changed to `auditLog`
- A `status` field must be provided to all `auditLog` function calls
- A `level: 'error'` field should be provided to all `auditLog` calls converted from `auditErrorLog`
@AndrienkoAleksandr what is your opinion on this change? Would it severely break your RBAC audit logging?
